apps/rebar/src/rebar_prv_completion.erl

```diff
@@ -0,0 +1,165 @@
+do(State) ->
+    DefaultOpts = #{aliases => [],
+                    file => "_rebar3",
+                    shell => detect_shell()},
+    {CliOptsList, _} = rebar_state:command_parsed_args(State),
+    CliOpts = maps:from_list(CliOptsList),
+    Conf = maps:from_list(rebar_state:get(State, completion, [])),
+    %% Opts passed in CLI override config
+    CmplOpts0 = maps:merge(DefaultOpts, Conf),
+    CmplOpts = maps:merge(CmplOpts0, CliOpts),
+
+    Providers0 = rebar_state:providers(State),
+    BareProviders = lists:filter(fun(P) -> provider_get(P, bare) end, Providers0),
+    ByNamespace = maps:groups_from_list(fun(P) -> provider_get(P, namespace) end, BareProviders),
+    Cmds0 = maps:fold(
+                fun(NS,Ps,CmdAcc) -> namespace_to_cmpl_cmds(NS, Ps)++CmdAcc end,
+                [],
+                ByNamespace),
+    Cmds = [oracle(Cmd, CmplOpts, State) || Cmd <- Cmds0],
+    Compl = rebar_completion:generate(Cmds, CmplOpts),
+    write_completion(Compl,State,CmplOpts),
+    {ok, State}.
+
+detect_shell() ->
+    case os:getenv("SHELL") of
+        false ->
+            ?DIAGNOSTIC("SHELL variable not set, default shell will be used.",
+                        []),
+            ?DEF_SHELL;
+        Path  ->
+            to_shell(filename:basename(Path))
+    end.
+
+to_shell("bash") -> bash;
+to_shell("zsh") -> zsh;
+to_shell(Unsupp) ->
+        ?WARN("Unsupported shell found: ~p, default shell will be used.",
+                    [Unsupp]),
+        ?DEF_SHELL.
+
+-spec namespace_to_cmpl_cmds(atom(), [providers:t()]) -> [rebar_completion:cmpl_cmd()].
+namespace_to_cmpl_cmds(default,Providers) ->
+    lists:map(fun(P)->provider_to_cmpl_cmd(P) end,Providers);
+namespace_to_cmpl_cmds(Namespace,Providers) ->
+    Name = atom_to_list(Namespace),
+    [#{name=>Name,
+      commands=>lists:map(fun(P)->provider_to_cmpl_cmd(P) end, Providers),
+      arguments=>[],
+      help=>Name++" namespace"}].
+
+-spec provider_to_cmpl_cmd(providers:t()) -> rebar_completion:cmpl_cmd().
+provider_to_cmpl_cmd(Provider) ->
+    Opts = providers:opts(Provider),
+    Name = providers:impl(Provider),
+    Cmd = getopt_to_cmpl_cmd(atom_to_list(Name),Opts),
+    Help = provider_get(Provider, short_desc),
+    Cmd#{help=>Help}.
+
+-spec getopt_to_cmpl_cmd(string(), [tuple()]) -> rebar_completion:cmpl_cmd().
+getopt_to_cmpl_cmd(Name, Opts) ->
+    Args = [#{short=>S,
+            long=>L,
+            type=>cmpl_arg_type(Spec),
+            help=>H} || {_,S,L,Spec,H} <- Opts],
+    #{name => Name,
+    arguments => Args,
+    commands => [],
+    help => undefined}.
+
+cmpl_arg_type({Type,_Default}) ->
+    Type;
+cmpl_arg_type(Type) ->
+    Type.
+
+%% ad-hoc injection of data for some known providers!
+-spec oracle(rebar_completion:cmpl_cmd(),
+             rebar_completion:cmpl_opts(),
+             rebar_state:t()) -> rebar_completion:cmpl_cmd().
+oracle(#{name:="as"}=Cmd, _CmplOpts, State) ->
```

Comment:
I imagine `do` is going to be similar here.

Reply:
Yes, **but* one uses `as` usually with one command, while `do` makes sense only for list of comma-separated values. Need to think about how to add comma-separated stuff. We could probably have logic like `if previous word ends with comma, trigger autocompletion again`, but that would require changes in completion file.

Reply:
Right. Either way though, not a blocker. Getting some autocompletion is already an improvement on none.

Reply:
I resolve it in another (more elegant I believe) way. Now when you type: `rebar3 do CMD <tab>` you have the same completion as you'd have if you wrote: `rebar CMD <tab>`, that means you have namespaced commands available. I might implement support for comma separated values some day.